csharp/src/Ice/InputStream.cs

```diff
@@ -1327,22 +1442,103 @@ public void Dispose()
+        // on-demand. It does not fully implement IEnumerable<T> and ICollection<T> (i.e. some methods throw
+        // NotSupportedException) because it's not resettable: you can't use it to unmarshal the same bytes multiple
+        // times.
+        private sealed class Collection<T> : CollectionBase<T>
+        {
+            private readonly InputStreamReader<T> _reader;
+
+            internal Collection(InputStream istr, int minElementSize, InputStreamReader<T> reader)
+                : base(istr, minElementSize) => _reader = reader;
+
+            public override bool MoveNext()
+            {
+                if (_pos < Count)
                 {
-                    _pos = Count + 1;
-                    return false;
+                    _current = _reader(_inputStream);
+                    _pos++;
+                    return true;
                 }
                 else
                 {
-                    _current = _reader(_inputStream);
+                    return false;
+                }
+            }
+        }
+
+        // Similar to Collection<T>, except we are reading a sequence<T?> where T is a reference type. T here must not
+        // correspond to a mapped Slice class or to a proxy class.
+        private sealed class OptionalElementCollection<T> : CollectionBase<T?> where T : class
+        {
+            private readonly ReadOnlyMemory<byte> _bitSequenceMemory;
+            private readonly InputStreamReader<T> _reader;
+
+            internal OptionalElementCollection(InputStream istr, InputStreamReader<T> reader)
+                : base(istr, 0)
+            {
+                _bitSequenceMemory = istr.ReadBitSequenceMemory(Count);
+                _reader = reader;
+            }
+
+            public override bool MoveNext()
+            {
+                if (_pos < Count)
+                {
+                    var bitSequence = new ReadOnlyBitSequence(_bitSequenceMemory.Span);
+                    _current = bitSequence[_pos++] ? _reader(_inputStream) : null;
                     return true;
                 }
+                else
+                {
+                    return false;
+                }
             }
+        }
 
-            public bool Remove(T item) => throw new NotSupportedException();
-            public void Reset() => throw new NotSupportedException();
+        // Similar to Collection<T>, except we are reading a sequence<T?> where T is a value type.
+        private sealed class OptionalValueCollection<T> : CollectionBase<T?> where T : struct
+        {
+            private readonly ReadOnlyMemory<byte> _bitSequenceMemory;
+            private readonly InputStreamReader<T> _reader;
+
+            internal OptionalValueCollection(InputStream istr, InputStreamReader<T> reader)
+                : base(istr, 0)
+            {
+                _bitSequenceMemory = istr.ReadBitSequenceMemory(Count);
```

Comment:
Why don't we just cache the ReadOnlyBitSequence instead? Does BitSequence really needs to be a ref struct? It would be good to add some comments to explain why.

Reply:
`BitSequence` holds one or two spans. Span<T> is a ref struct, so `BitSequence` needs to be a ref struct. Same for `ReadOnlyBitSequence`. I thought this was reasonably obvious ... I'll add a comment.

Reply:
And we can't just use ReadOnlyMemory to back the ReadOnlyBitSequence instead of a span?

Reply:
We could, but this would be inconsistent with `BitSequence`. I don't think the current implementation which creates a `ReadOnlyBitSequence` "on the fly" for each `MoveNext` is particularly slow as a `ReadOnlyBitSequence` is a lightweight stack-allocated ref struct.